### ISSUE TITLE
JWK caching, no handle on configuring refresh rate of public keys for a provider. 

### ISSUE THREAD

**Issue by gumond** [OpenIDC/pyoidc#702]
pyoidc library caches the JWKs for 5 minutes. After that, if one calls again the client.keyjar.get_issuer_keys, the library will make a new request to the provider. 
Is there or will there be any way to control the caching time to have better handle 
on how often the "offline validation" happens ?

**Comment by tpazderka** [comment]
This is possible in the current state (although not very straightforward...).

Client takes a Keyjar instance to which a KeyJar instance can be added directly. KeyJar instance then takes a cache_timeout parameter (in seconds).

So getting the key separately and then adding it to the KeyJar should allow you to set a longer caching time.

We are open to suggestions (and pull requests) so if you have something different in mind, feel free to let us know.

**Comment by gumond** [comment]
This is already very helpful, Thank you for this fast feedback ! I will try to figure out an example on how to do what you describe.

**Comment by gumond** [comment]
well .. trying to follow, I see that there is only 'cache_time' parameter which is a parameter of KeyBundle rather than KeyJar. Can I add KeyBundle to the Client instead ? Would you have an example ?

**Comment by tpazderka** [comment]
No, you cannot. But you can add KeyBundle to a KeyJar (using `add_kb`) and that can be passed to Client.

**Comment by gumond** [comment]
And what about just setting the value for already initializes client ? Something like this: 
client.keyjar.issuer_keys[client.issuer][0].cache_time = 3000
?

**Comment by tpazderka** [comment]
A little bit of a hack, but should in theory work.

**Comment by gumond** [comment]
Well I tested it (looping through the keybundle list) and asking for jwk from the IdP and it does work.  Will this 'hack' be always possible or do you plan to prevent such access ? If the latter, I imagine we would need to allow such parameters (maybe there is more of them) to be specified easily at client creation time. I do not think that using 'internal' objects (like KeyJar, KeyBundle, etc.) of the pyoidc library is what I wish to do as I am happy that the library does these things for me.

**Comment by tpazderka** [comment]
Glad to hear that!

There is no plan to remove this functionality as of now, but it is sort of a hack and as such can stop working :/ But it should work as the dict-like interface will most likely stay.

Adding some other configuration options to the Client/Server classes is a little bit complicated as their init methods are already a mess and the configuration options for client are not well documented.

I have a redesign in mind, that would make adding new configuration options easier, but it will need some time :)
